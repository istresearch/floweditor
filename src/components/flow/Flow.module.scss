@import 'variables.module.scss';

.empty_flow {
  margin: 30px;
  width: 400px;

  h1 {
    font-family: 'Roboto', sans-serif;
    letter-spacing: 1px;
    padding: 0;
    text-transform: none;
    font-weight: 100;
    font-size: 18px;
    margin: 0;
  }

  div {
    margin-bottom: 20px;
    font-size: 14px;
    line-height: 18px;
  }
}