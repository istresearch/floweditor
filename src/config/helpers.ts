import {
  FlowTypes,
  FlowTypeVisibility,
  Operator,
  Type,
  VISIBILITY_ONLINE
} from 'config/interfaces';
import { FlowEditorConfig } from 'flowTypes';

export const isOnlineFlowType = (flowType: FlowTypes) => {
  return !!VISIBILITY_ONLINE.find((type: FlowTypes) => type === flowType);
};

export const filterOperators = (operators: Operator[], config: FlowEditorConfig): Operator[] => {
  return filterVisibility(operators, config);
};

export const filterTypeConfigs = (typeConfigs: Type[], config: FlowEditorConfig): Type[] => {
  return filterVisibility(typeConfigs, config);
};

const filterVisibility = (items: FlowTypeVisibility[], config: FlowEditorConfig): any[] => {
  return items.filter((item: FlowTypeVisibility) => {
    // if we have a filter on our type, don't return it unless its present in our config
    if (item.filter) {
      if (!(config.filters || []).find((name: string) => name === item.filter)) {
        return false;
      }
    }

    if (item.visibility === undefined) {
      return true;
    }

    return item.visibility.findIndex((ft: FlowTypes) => ft === config.flowType) > -1;
  });
};
