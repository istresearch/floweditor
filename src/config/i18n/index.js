import i18n from 'i18next';
import { initReactI18next } from 'react-i18next';
import LanguageDetector from 'i18next-browser-languagedetector';

// TODO: consider chunking out translation files
// and using something like XHR backend to lazily
// load only the required language.
// See https://www.i18next.com/how-to/add-or-load-translations

import defaults from './defaults.json';
import es from './es/resource.json';
import fr from './fr/resource.json';
import ptBR from './pt-br/resource.json';

i18n
  .use(LanguageDetector)
  .use(initReactI18next)
  .init({
    detection: {
      order: ['querystring', 'htmlTag']
    },
    resources: {
      defaults: { translation: defaults },
      es: { translation: es },
      fr: { translation: fr },
      'pt-br': { translation: ptBR }
    },
    fallbackLng: ['defaults'],
    interpolation: {
      prefix: '[[',
      suffix: ']]'
    },
    lowerCaseLng: true
  });

export default i18n;
