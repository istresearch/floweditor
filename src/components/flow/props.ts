import { Type } from 'config/interfaces';
import { AnyAction, ContactProperties, FlowIssue } from 'flowTypes';
import { Asset, AssetStore, AssetType, RenderNode } from 'store/flowContext';
import { NodeEditorSettings } from 'store/nodeEditor';
import { DispatchWithState, GetState } from 'store/thunks';
import { titleCase } from 'utils';

export interface IssueProps {
  helpArticles: { [key: string]: string };
  issues: FlowIssue[];
}

export interface ActionFormProps extends IssueProps {
  // action details
  nodeSettings: NodeEditorSettings;
  typeConfig: Type;
  assetStore: AssetStore;

  addAsset(assetType: string, asset: Asset): void;

  // update handlers
  updateAction(
    action: AnyAction,
    onUpdated?: (dispatch: DispatchWithState, getState: GetState) => void
  ): void;

  // modal notifiers
  onTypeChange(config: Type): void;
  onClose(canceled: boolean): void;
}

export interface RouterFormProps extends IssueProps {
  nodeSettings: NodeEditorSettings;
  typeConfig: Type;

  assetStore: AssetStore;

  // update handlers
  updateRouter(renderNode: RenderNode): void;

  // modal notifiers
  onTypeChange(config: Type): void;
  onClose(canceled: boolean): void;
}

export interface LocalizationFormProps extends IssueProps {
  language: Asset;
  nodeSettings: NodeEditorSettings;
  updateLocalizations(languageCode: string, localizations: any[]): void;
  onClose(canceled: boolean): void;
  helpArticles: { [key: string]: string };
}

export const NAME_PROPERTY: Asset = {
  name: titleCase(ContactProperties.Name),
  id: ContactProperties.Name,
  type: AssetType.ContactProperty
};

export const CHANNEL_PROPERTY: Asset = {
  name: titleCase(ContactProperties.Channel),
  id: ContactProperties.Channel,
  type: AssetType.ContactProperty
};

export const LANGUAGE_PROPERTY: Asset = {
  name: titleCase(ContactProperties.Language),
  id: ContactProperties.Language,
  type: AssetType.ContactProperty
};

export const STATUS_PROPERTY: Asset = {
  name: titleCase(ContactProperties.Status),
  id: ContactProperties.Status,
  type: AssetType.ContactProperty
};
