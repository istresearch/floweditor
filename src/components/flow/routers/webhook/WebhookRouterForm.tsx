import { react as bindCallbacks } from 'auto-bind';
import Dialog, { ButtonSet, Tab } from 'components/dialog/Dialog';
import { hasErrors, renderIssues } from 'components/flow/actions/helpers';
import { RouterFormProps } from 'components/flow/props';
import HeaderElement, { Header } from 'components/flow/routers/webhook/header/HeaderElement';
import {
  METHOD_OPTIONS,
  MethodOption,
  Methods,
  nodeToState,
  stateToNode,
  getDefaultBody
} from 'components/flow/routers/webhook/helpers';
import { createResultNameInput } from 'components/flow/routers/widgets';
import SelectElement from 'components/form/select/SelectElement';
import TextInputElement from 'components/form/textinput/TextInputElement';
import TypeList from 'components/nodeeditor/TypeList';
import * as React from 'react';
import { FormEntry, FormState, mergeForm, StringEntry, ValidationFailure } from 'store/nodeEditor';
import {
  Alphanumeric,
  shouldRequireIf,
  StartIsNonNumeric,
  validate,
  ValidURL,
  validateIf
} from 'store/validators';
import { createUUID } from 'utils';

import styles from './WebhookRouterForm.module.scss';
import { Trans } from 'react-i18next';
import i18n from 'config/i18n';

export interface HeaderEntry extends FormEntry {
  value: Header;
}

export interface MethodEntry extends FormEntry {
  value: MethodOption;
}

export interface WebhookRouterFormState extends FormState {
  headers: HeaderEntry[];
  method: MethodEntry;
  url: StringEntry;
  body: StringEntry;
  resultName: StringEntry;
}

export default class WebhookRouterForm extends React.Component<
  RouterFormProps,
  WebhookRouterFormState
> {
  constructor(props: RouterFormProps) {
    super(props);
    this.state = nodeToState(this.props.nodeSettings);
    bindCallbacks(this, {
      include: [/^handle/]
    });
  }

  private handleUpdate(
    keys: {
      method?: MethodOption;
      url?: string;
      body?: string;
      header?: Header;
      removeHeader?: Header;
      validationFailures?: ValidationFailure[];
      resultName?: string;
    },
    submitting = false
  ): boolean {
    const updates: Partial<WebhookRouterFormState> = {};

    let ensureEmptyHeader = false;
    let toRemove: any[] = [];

    if (keys.hasOwnProperty('method')) {
      updates.method = { value: keys.method };

      const oldMethod = this.state.method.value.value;
      const newMethod = keys.method.value;

      if (oldMethod !== newMethod) {
        const existingContentTypeHeader = this.state.headers.find(
          (header: HeaderEntry) => header.value.name.toLowerCase() === 'content-type'
        );

        // whenever our method changes, update the default body
        updates.body = { value: getDefaultBody(newMethod) };

        // switching from a GET, add a content-type
        if (oldMethod === Methods.GET && newMethod !== Methods.GET) {
          if (!existingContentTypeHeader) {
            let uuid = createUUID();
            // if we have an empty header, use that one
            const lastHeader =
              this.state.headers.length > 0
                ? this.state.headers[this.state.headers.length - 1]
                : null;
            if (lastHeader && !lastHeader.value.name) {
              uuid = lastHeader.value.uuid;
            }
            keys.header = { uuid, name: 'Content-Type', value: 'application/json' };
          }
        } else if (oldMethod !== Methods.GET && newMethod === Methods.GET) {
          // remove content type if switching to a GET
          if (existingContentTypeHeader) {
            toRemove = [{ headers: [{ value: existingContentTypeHeader.value }] }];
          }
        }
      }
    }

    if (keys.hasOwnProperty('url')) {
      updates.url = validate(i18n.t('forms.url', 'URL'), keys.url, [
        shouldRequireIf(submitting),
        validateIf(ValidURL, keys.url.indexOf('@') === -1)
      ]);
    }

    if (keys.hasOwnProperty('resultName')) {
      updates.resultName = validate(i18n.t('forms.result_name', 'Result Name'), keys.resultName, [
        shouldRequireIf(submitting),
        Alphanumeric,
        StartIsNonNumeric
      ]);
    }

    if (keys.hasOwnProperty('body')) {
      updates.body = { value: keys.body };
    }

    if (keys.hasOwnProperty('header')) {
      updates.headers = [{ value: keys.header, validationFailures: keys.validationFailures }];
      ensureEmptyHeader = true;
    }

    if (keys.hasOwnProperty('removeHeader')) {
      toRemove = [{ headers: [{ value: keys.removeHeader }] }];
      ensureEmptyHeader = true;
    }

    const updated = mergeForm(this.state, updates, toRemove);

    // update our form
    this.setState(updated, () => {
      // if we updated headers, check if we need a new one
      if (ensureEmptyHeader) {
        let needsHeader = true;
        for (const header of this.state.headers) {
          if (header.value.name.trim() === '') {
            needsHeader = false;
            break;
          }
        }

        if (needsHeader) {
          this.handleCreateHeader();
        }
      }
    });
    return updated.valid;
  }

  private handleUpdateResultName(value: string): boolean {
    return this.handleUpdate({ resultName: value });
  }

  private handleMethodUpdate(method: MethodOption): boolean {
    return this.handleUpdate({ method });
  }

  private handleUrlUpdate(url: string, name: string, submitting = false): boolean {
    return this.handleUpdate({ url }, submitting);
  }

  private handleHeaderRemoved(removeHeader: Header): boolean {
    return this.handleUpdate({ removeHeader });
  }

  private handleHeaderUpdated(header: Header, validationFailures: ValidationFailure[]): boolean {
    return this.handleUpdate({ header, validationFailures });
  }

  private handleCreateHeader(): boolean {
    return this.handleUpdate({
      header: {
        uuid: createUUID(),
        name: '',
        value: ''
      }
    });
  }

  private handleBodyUpdate(body: string): boolean {
    return this.handleUpdate({ body });
  }

  private handleSave(): void {
    // validate our url in case they haven't interacted
    const valid = this.handleUpdate(
      { url: this.state.url.value, resultName: this.state.resultName.value },
      true
    );

    if (valid) {
      this.props.updateRouter(stateToNode(this.props.nodeSettings, this.state));
      this.props.onClose(false);
    }
  }

  private getButtons(): ButtonSet {
    return {
      primary: { name: i18n.t('buttons.ok', 'Ok'), onClick: this.handleSave },
      secondary: {
        name: i18n.t('buttons.cancel', 'Cancel'),
        onClick: () => this.props.onClose(true)
      }
    };
  }

  private renderEdit(): JSX.Element {
    const typeConfig = this.props.typeConfig;

    const headerElements: JSX.Element[] = this.state.headers.map(
      (header: HeaderEntry, index: number, arr: HeaderEntry[]) => {
        return (
          <div key={`header_${header.value.uuid}`}>
            <HeaderElement
              entry={header}
              onRemove={this.handleHeaderRemoved}
              onChange={this.handleHeaderUpdated}
              index={index}
            />
          </div>
        );
      }
    );

    const tabs: Tab[] = [];
    tabs.push({
      name: i18n.t('forms.http_headers', 'HTTP Headers'),
      hasErrors: !!this.state.headers.find((header: HeaderEntry) => hasErrors(header)),
      body: (
        <>
          <p className={styles.info}>
            <Trans i18nKey="forms.webhook_header_summary">
              Add any additional headers below that you would like to send along with your request.
            </Trans>
          </p>
          {headerElements}
        </>
      ),
      checked: this.state.headers.length > 1
    });

    const method = this.state.method.value.value;
    const name = this.state.method.value.name + ' ' + i18n.t('body', 'Body');
    tabs.push({
      name,
      body: (
        <div key="post_body" className={styles.body_form}>
          <h4>{name}</h4>
          <p>
            <Trans
              i18nKey="forms.webhook_body_summary"
              values={{ method: this.state.method.value.name }}
            >
              Modify the body of the [[method]] request that will be sent to your webhook.
            </Trans>
          </p>
          <TextInputElement
            __className={styles.req_body}
            name={name}
            showLabel={false}
            entry={this.state.body}
            onChange={this.handleBodyUpdate}
            helpText={
              <Trans
                i18nKey="forms.webhook_body_summary"
                values={{ method: this.state.method.value.name }}
              >
                Modify the body of the [[method]] request that will be sent to your webhook.
              </Trans>
            }
            autocomplete={true}
            textarea={true}
          />
        </div>
      ),
      checked: this.state.body.value !== getDefaultBody(method)
    });

    return (
      <Dialog
        title={typeConfig.name}
        headerClass={typeConfig.type}
        buttons={this.getButtons()}
        tabs={tabs}
      >
        <TypeList __className="" initialType={typeConfig} onChange={this.props.onTypeChange} />
        <div style={{ display: 'flex', alignItems: 'flex-end' }}>
          <div className={styles.method}>
            <SelectElement
              key="method_select"
              name={i18n.t('forms.method', 'Method')}
              entry={this.state.method}
              onChange={this.handleMethodUpdate}
              options={METHOD_OPTIONS}
            />
          </div>
          <div className={styles.url}>
            <TextInputElement
              name={i18n.t('forms.url', 'URL')}
              placeholder={i18n.t('forms.enter_a_url', 'Enter a URL')}
              entry={this.state.url}
              onChange={this.handleUrlUpdate}
              autocomplete={true}
            />
          </div>
        </div>
        <div className={styles.instructions}>
          <p>
            <Trans i18nKey="forms.webhook_help">
              If your server responds with JSON, each property will be added to the Flow.
            </Trans>
          </p>
          <pre className={styles.code}>
            {'{ "product": "Solar Charging Kit", "stock level": 32 }'}
          </pre>
          <p>
            <Trans i18nKey="forms.webhook_example">
              This response would add <span className={styles.example}>@webhook.product</span> and{' '}
              <span className={styles.example}>@webhook["stock level"]</span> for use in the flow.
            </Trans>
          </p>
        </div>
        {createResultNameInput(this.state.resultName, this.handleUpdateResultName)}
        {renderIssues(this.props)}
      </Dialog>
    );
  }

  public render(): JSX.Element {
    return this.renderEdit();
  }
}
