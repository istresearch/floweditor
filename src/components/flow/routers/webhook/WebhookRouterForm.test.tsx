/* eslint-disable @typescript-eslint/no-object-literal-type-assertion */
/* eslint-disable @typescript-eslint/explicit-function-return-type */
import { RouterFormProps } from 'components/flow/props';
import WebhookRouterForm from 'components/flow/routers/webhook/WebhookRouterForm';
import { Types } from 'config/interfaces';
import { RenderNode } from 'store/flowContext';
import { composeComponentTestUtils, mock } from 'testUtils';
import { createWebhookRouterNode, getRouterFormProps } from 'testUtils/assetCreators';
import * as utils from 'utils';
import * as React from 'react';
import { render, fireEvent, fireChangeText, fireTembaSelect } from 'test/utils';

mock(utils, 'createUUID', utils.seededUUIDs());

const webhookForm = getRouterFormProps({
  node: createWebhookRouterNode(),
  ui: { type: Types.call_webhook }
} as RenderNode);

const { setup } = composeComponentTestUtils<RouterFormProps>(WebhookRouterForm, webhookForm);

describe(WebhookRouterForm.name, () => {
  it('should render', () => {
    const { wrapper } = setup(true);
    expect(wrapper).toMatchSnapshot();
  });

  describe('updates', () => {
    it('should save changes', () => {
      const { baseElement, getByText, getByTestId, getAllByTestId } = render(
        <WebhookRouterForm {...webhookForm} />
      );
      expect(baseElement).toMatchSnapshot();

      const okButton = getByText('Ok');

      // our url is required
      fireEvent.click(okButton);
      expect(webhookForm.updateRouter).not.toBeCalled();

      // set our url and name
      const url = getByTestId('URL');
      const resultName = getByTestId('Result Name');

      fireChangeText(url, 'http://app.rapidpro.io');
      fireChangeText(resultName, 'My Webhook Result');

      // make it a post
      fireTembaSelect(getByTestId('temba_select_method'), 'POST');

      // set a post body
      fireEvent.click(getByText('POST Body'));
      const postBody = getByTestId('POST Body');

      fireChangeText(postBody, 'Updated post body');

      // add http header
      fireEvent.click(getByText('HTTP Headers'));
      const headerName = getAllByTestId('Header Name')[0];
      const headerValue = getAllByTestId('Value')[0];

      fireEvent.change(headerName, 'Content-type');
      fireEvent.change(headerValue, 'application/json');

      fireEvent.click(okButton);
      expect(webhookForm.updateRouter).toBeCalled();
      expect(webhookForm.updateRouter).toMatchCallSnapshot();
    });

    it('should repopulate post body', () => {
      const { instance } = setup(true, {
        $merge: { onClose: jest.fn(), updateRouter: jest.fn() }
      });

      instance.handleMethodUpdate({ value: 'GET' });
      instance.handleUrlUpdate('http://domain.com/');
      expect(instance.state).toMatchSnapshot();

      instance.handleMethodUpdate('POST');
      expect(instance.state).toMatchSnapshot();
    });

    it('should cancel', () => {
      const { instance, props } = setup(true, {
        $merge: { onClose: jest.fn(), updateRouter: jest.fn() }
      });
      instance.getButtons().secondary.onClick();
      instance.handleUrlUpdate('http://domain.com/');
      expect(props.onClose).toHaveBeenCalled();
      expect(props.updateRouter).not.toHaveBeenCalled();
    });

    it('should validate urls', () => {
      webhookForm.updateRouter = jest.fn();
      const { getByText, getByTestId } = render(<WebhookRouterForm {...webhookForm} />);

      // set our url and name
      const url = getByTestId('URL');
      const resultName = getByTestId('Result Name');

      fireChangeText(url, 'bad url');
      fireChangeText(resultName, 'My Webhook Result');

      // we need a valid url
      const okButton = getByText('Ok');
      fireEvent.click(okButton);
      expect(webhookForm.updateRouter).not.toBeCalled();

      // but not if it has an expression
      fireChangeText(url, '@fields.valid_url');
      fireEvent.click(okButton);
      expect(webhookForm.updateRouter).toBeCalled();
    });
  });
});
