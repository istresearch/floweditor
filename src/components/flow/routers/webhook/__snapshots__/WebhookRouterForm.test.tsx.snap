// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WebhookRouterForm should render 1`] = `
<Dialog
  buttons={
    Object {
      "primary": Object {
        "name": "Ok",
        "onClick": [Function],
      },
      "secondary": Object {
        "name": "Cancel",
        "onClick": [Function],
      },
    }
  }
  headerClass="call_webhook"
  tabs={
    Array [
      Object {
        "body": <React.Fragment>
          <p
            className="info"
          >
            <Trans
              i18nKey="forms.webhook_header_summary"
            >
              Add any additional headers below that you would like to send along with your request.
            </Trans>
          </p>
          <div>
            <HeaderElement
              entry={
                Object {
                  "value": Object {
                    "name": "Accept",
                    "uuid": "2dc85899-0577-4f1b-a620-f12094e34b5e",
                    "value": "application/json",
                  },
                }
              }
              index={0}
              onChange={[Function]}
              onRemove={[Function]}
            />
          </div>
          <div>
            <HeaderElement
              entry={
                Object {
                  "value": Object {
                    "name": "",
                    "uuid": "42ebf1d0-a43e-4ba6-9577-c863b20c88c4",
                    "value": "",
                  },
                }
              }
              index={1}
              onChange={[Function]}
              onRemove={[Function]}
            />
          </div>
        </React.Fragment>,
        "checked": true,
        "hasErrors": false,
        "name": "HTTP Headers",
      },
      Object {
        "body": <div
          className="body_form"
        >
          <h4>
            GET Body
          </h4>
          <p>
            <Trans
              i18nKey="forms.webhook_body_summary"
              values={
                Object {
                  "method": "GET",
                }
              }
            >
              Modify the body of the [[method]] request that will be sent to your webhook.
            </Trans>
          </p>
          <TextInputElement
            __className="req_body"
            autocomplete={true}
            entry={
              Object {
                "value": "",
              }
            }
            helpText={
              <Trans
                i18nKey="forms.webhook_body_summary"
                values={
                  Object {
                    "method": "GET",
                  }
                }
              >
                Modify the body of the [[method]] request that will be sent to your webhook.
              </Trans>
            }
            name="GET Body"
            onChange={[Function]}
            showLabel={false}
            textarea={true}
          />
        </div>,
        "checked": false,
        "name": "GET Body",
      },
    ]
  }
  title="Call Webhook"
>
  <TypeList
    __className=""
    initialType={
      Object {
        "aliases": Array [
          "split_by_webhook",
        ],
        "component": [Function],
        "description": "Call a webhook",
        "form": [Function],
        "localization": [Function],
        "localizeableKeys": Array [
          "exits",
        ],
        "name": "Call Webhook",
        "type": "call_webhook",
        "visibility": Array [
          "messaging",
          "messaging_background",
          "voice",
        ],
      }
    }
    onChange={[MockFunction]}
  />
  <div
    style={
      Object {
        "alignItems": "flex-end",
        "display": "flex",
      }
    }
  >
    <div
      className="method"
    >
      <SelectElement
        entry={
          Object {
            "value": Object {
              "name": "GET",
              "value": "GET",
            },
          }
        }
        key="method_select"
        name="Method"
        onChange={[Function]}
        options={
          Array [
            Object {
              "name": "GET",
              "value": "GET",
            },
            Object {
              "name": "POST",
              "value": "POST",
            },
            Object {
              "name": "PUT",
              "value": "PUT",
            },
            Object {
              "name": "DELETE",
              "value": "DELETE",
            },
            Object {
              "name": "HEAD",
              "value": "HEAD",
            },
            Object {
              "name": "PATCH",
              "value": "PATCH",
            },
          ]
        }
      />
    </div>
    <div
      className="url"
    >
      <TextInputElement
        autocomplete={true}
        entry={
          Object {
            "value": "",
          }
        }
        name="URL"
        onChange={[Function]}
        placeholder="Enter a URL"
      />
    </div>
  </div>
  <div
    className="instructions"
  >
    <p>
      <Trans
        i18nKey="forms.webhook_help"
      >
        If your server responds with JSON, each property will be added to the Flow.
      </Trans>
    </p>
    <pre
      className="code"
    >
      { "product": "Solar Charging Kit", "stock level": 32 }
    </pre>
    <p>
      <Trans
        i18nKey="forms.webhook_example"
      >
        This response would add 
        <span
          className="example"
        >
          @webhook.product
        </span>
         and
         
        <span
          className="example"
        >
          @webhook["stock level"]
        </span>
         for use in the flow.
      </Trans>
    </p>
  </div>
  <OptionalTextInput
    helpText={
      <Trans
        i18nKey="forms.result_name_help"
        values={
          Object {
            "resultFormat": "@results.result",
          }
        }
      >
        By naming the result, you can reference it later using [[resultFormat]]
      </Trans>
    }
    maxLength={64}
    name="Result Name"
    onChange={[Function]}
    toggleText="Save as.."
    value={
      Object {
        "value": "Result",
      }
    }
  />
</Dialog>
`;

exports[`WebhookRouterForm updates should repopulate post body 1`] = `
Object {
  "body": Object {
    "value": "",
  },
  "headers": Array [
    Object {
      "value": Object {
        "name": "Accept",
        "uuid": "0c8c9239-9b59-473a-a229-d8a9693be270",
        "value": "application/json",
      },
    },
    Object {
      "value": Object {
        "name": "",
        "uuid": "9adce29f-da95-4182-9cb3-0f1e3bf714e5",
        "value": "",
      },
    },
  ],
  "method": Object {
    "value": Object {
      "value": "GET",
    },
  },
  "resultName": Object {
    "value": "Result",
  },
  "url": Object {
    "validationFailures": Array [],
    "value": "http://domain.com/",
  },
  "valid": true,
}
`;

exports[`WebhookRouterForm updates should repopulate post body 2`] = `
Object {
  "body": Object {
    "value": "@(json(object(
  \\"contact\\", object(
    \\"uuid\\", contact.uuid, 
    \\"name\\", contact.name, 
    \\"urn\\", contact.urn
  ),
  \\"flow\\", object(
    \\"uuid\\", run.flow.uuid, 
    \\"name\\", run.flow.name
  ),
  \\"results\\", foreach_value(results, extract_object, \\"value\\", \\"category\\")
)))",
  },
  "headers": Array [
    Object {
      "value": Object {
        "name": "Accept",
        "uuid": "0c8c9239-9b59-473a-a229-d8a9693be270",
        "value": "application/json",
      },
    },
    Object {
      "validationFailures": undefined,
      "value": Object {
        "name": "Content-Type",
        "uuid": "9adce29f-da95-4182-9cb3-0f1e3bf714e5",
        "value": "application/json",
      },
    },
    Object {
      "validationFailures": undefined,
      "value": Object {
        "name": "",
        "uuid": "00dd2423-0a34-4ee2-b97d-839fb2d593ef",
        "value": "",
      },
    },
  ],
  "method": Object {
    "value": "POST",
  },
  "resultName": Object {
    "value": "Result",
  },
  "url": Object {
    "validationFailures": Array [],
    "value": "http://domain.com/",
  },
  "valid": true,
}
`;

exports[`WebhookRouterForm updates should save changes 1`] = `
<body>
  <div>
    <div
      class="dialog"
    >
      <div
        class="tabs"
      >
        <div
          class="tab "
        >
          HTTP Headers
           
          <span
            class="tab_icon fe-check"
          />
        </div>
        <div
          class="tab "
        >
          GET Body
           
        </div>
      </div>
      <div
        class="header call_webhook"
      >
        <div
          class="title_container"
        >
          <div
            class="title"
          >
            Call Webhook
          </div>
          <div
            class="subtitle"
          />
        </div>
      </div>
      <div
        class="content"
      >
        <div
          class=" type_list"
        >
          <p>
            When a contact arrives at this point in your flow...
          </p>
          <div>
            <div
              class="normal "
            >
              <temba-select
                data-testid="temba_select_type"
                errors="[]"
                expressions=""
                name="type"
                namekey="description"
                searchable="true"
                valuekey="type"
                values="[{\\"type\\":\\"call_webhook\\",\\"name\\":\\"Call Webhook\\",\\"description\\":\\"Call a webhook\\",\\"localizeableKeys\\":[\\"exits\\"],\\"aliases\\":[\\"split_by_webhook\\"],\\"visibility\\":[\\"messaging\\",\\"messaging_background\\",\\"voice\\"]}]"
              />
            </div>
          </div>
        </div>
        <div
          style="display: flex; align-items: flex-end;"
        >
          <div
            class="method"
          >
            <div
              class="ele"
            >
              <div
                class="normal "
              >
                <temba-select
                  data-testid="temba_select_method"
                  errors="[]"
                  expressions=""
                  name="Method"
                  namekey="name"
                  valuekey="value"
                  values="[{\\"value\\":\\"GET\\",\\"name\\":\\"GET\\"}]"
                />
              </div>
            </div>
          </div>
          <div
            class="url"
          >
            <div
              class="ele"
            >
              <div
                class="wrapper normal"
              >
                <input
                  data-testid="URL"
                  name="URL"
                  placeholder="Enter a URL"
                  value=""
                />
              </div>
            </div>
          </div>
        </div>
        <div
          class="instructions"
        >
          <p>
            If your server responds with JSON, each property will be added to the Flow.
          </p>
          <pre
            class="code"
          >
            { "product": "Solar Charging Kit", "stock level": 32 }
          </pre>
          <p>
            This response would add 
            <span
              class="example"
            >
              @webhook.product
            </span>
             and 
            <span
              class="example"
            >
              @webhook["stock level"]
            </span>
             for use in the flow.
          </p>
        </div>
        <div
          class="optional_text_input"
        >
          <div
            class="ele"
          >
            <div
              class="label"
            >
              Result Name
            </div>
            <div
              class="wrapper normal"
            >
              <input
                data-testid="Result Name"
                name="Result Name"
                value="Result"
              />
            </div>
            <div
              class="help_text"
            >
              By naming the result, you can reference it later using @results.result
               
            </div>
          </div>
        </div>
      </div>
      <div
        class="footer"
      >
        <div
          class="buttons"
        >
          <div
            class="right_buttons"
          >
            <div
              class="btn secondary  "
              style="margin-left: 0px; margin-top: 0px;"
            >
              Cancel
            </div>
            <div
              class="btn primary  "
              style="margin-left: 10px; margin-top: 0px;"
            >
              Ok
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`WebhookRouterForm updates should save changes 2`] = `
Array [
  Object {
    "inboundConnections": Object {},
    "node": Object {
      "actions": Array [
        Object {
          "body": "Updated post body",
          "headers": Object {
            "Accept": "application/json",
            "Content-Type": "application/json",
          },
          "method": "POST",
          "result_name": "My Webhook Result",
          "type": "call_webhook",
          "url": "http://app.rapidpro.io",
          "uuid": "b1f332f3-bdd3-4891-aec5-1843a712dbf1",
        },
      ],
      "exits": Array [
        Object {
          "destination_uuid": null,
          "uuid": "061fc171-8b79-4636-b892-bd0ea5aa9b42",
        },
        Object {
          "destination_uuid": null,
          "uuid": "1dce2b34-9aab-4e20-81c4-3f0408dcb671",
        },
      ],
      "router": Object {
        "cases": Array [
          Object {
            "arguments": Array [
              "Success",
            ],
            "category_uuid": "763e4844-3e1b-407a-a1b5-5fdfcd308b41",
            "type": "has_only_text",
            "uuid": "1e47a1e1-3c67-4df5-adf1-da542c789adb",
          },
        ],
        "categories": Array [
          Object {
            "exit_uuid": "061fc171-8b79-4636-b892-bd0ea5aa9b42",
            "name": "Success",
            "uuid": "763e4844-3e1b-407a-a1b5-5fdfcd308b41",
          },
          Object {
            "exit_uuid": "1dce2b34-9aab-4e20-81c4-3f0408dcb671",
            "name": "Failure",
            "uuid": "477ac8b4-25e2-483a-8686-2d1332c4da1c",
          },
        ],
        "default_category_uuid": "477ac8b4-25e2-483a-8686-2d1332c4da1c",
        "operand": "@results.my_webhook_result.category",
        "type": "switch",
      },
      "uuid": "132de855-4042-4dc1-a18f-cc2e6a8f790a",
    },
    "ui": Object {
      "config": Object {},
      "position": null,
      "type": "split_by_webhook",
    },
  },
]
`;
